examples/2d/elixir_advection_p4est_non_conforming.jl

```diff
@@ -0,0 +1,83 @@
+
+using OrdinaryDiffEq
+using P4est
+using Trixi
+
+
+# Refine bottom left quadrant of each forest to level 4
+function refine_fn(p4est, which_tree, quadrant)
+  if quadrant.x == 0 && quadrant.y == 0 && quadrant.level < 4
+    # return true (refine)
+    return Cint(1)
+  else
+    # return false (don't refine)
+    return Cint(0)
+  end
+end
+
+###############################################################################
+# semidiscretization of the linear advection equation
+
+advectionvelocity = (1.0, 1.0)
+equations = LinearScalarAdvectionEquation2D(advectionvelocity)
+
+# Create DG solver with polynomial degree = 3 and (local) Lax-Friedrichs/Rusanov flux as surface flux
+solver = DGSEM(polydeg=3, surface_flux=flux_lax_friedrichs)
+
+# Deformed rectangle that looks like a waving flag,
+# lower and upper faces are sinus curves, left and right are vertical lines.
+f1(s) = SVector(-1.0, s - 1.0)
+f2(s) = SVector( 1.0, s + 1.0)
+f3(s) = SVector(s, -1.0 + sin(0.5 * pi * s))
+f4(s) = SVector(s,  1.0 + sin(0.5 * pi * s))
+
+trees_per_dimension = (3, 2)
+
+# Create P4estMesh with 8 x 8 trees and 16 x 16 elements
+mesh = P4estMesh(trees_per_dimension, polydeg=3,
+                 faces=(f1, f2, f3, f4),
+                 initial_refinement_level=1)
+
+# Refine recursively until each bottom left quadrant of a forest has level 4
+# The mesh will be rebalanced before the simulation starts
+refine_fn_c = @cfunction(refine_fn, Cint, (Ptr{p4est_t}, Ptr{p4est_topidx_t}, Ptr{p4est_quadrant_t}))
+p4est_refine(mesh.p4est, true, refine_fn_c, C_NULL)
```

Comment:
Minor bikeshedding that you can ignore: Since you don't use `refine_fn` anywhere else and call it only once, I would prefer to define it in a `let` block to avoid cluttering the global namespace.

Reply:
What would this look like?
And does it work with `@cfunction`? I noticed that I can't define the function after the function where I call `@cfunction`, and defining it inside this function doesn't work either.

Reply:
It could look like this.
```suggestion
let # We use a `let` block to avoid cluttering the global namespace.
  # Refine recursively until each bottom left quadrant of a forest has level 4.
  # The mesh will be rebalanced before the simulation starts.
  function refine_fn(p4est, which_tree, quadrant)
    if quadrant.x == 0 && quadrant.y == 0 && quadrant.level < 4
      # return true (refine)
      return Cint(1)
    else
      # return false (don't refine)
      return Cint(0)
    end
  end
  refine_fn_c = @cfunction($refine_fn, Cint,
    (Ptr{Trixi.p4est_t}, Ptr{Trixi.p4est_topidx_t}, Ptr{Trixi.p4est_quadrant_t}))

  GC.@preserve refine_fn_c begin
    # Don't let the GC collect `refine_fn_c` while p4est is working with it.
    Trixi.p4est_refine(mesh.p4est, true,
      Base.unsafe_convert(Ptr{Cvoid}, refine_fn_c), C_NULL)
  end
end
```
However, I don't know whether this is considered as "too advanced" and worth the effort. Thus, you can also use the current version, I guess. @sleode?

Nevertheless, I would prefer having the definition of `refine_fn` where it is needed instead of at the beginning of the file, as pointed out above.

Reply:
This will also prevent `invalid redefinition of function mapping` if one elixir defines `mapping = Trixi.somemapping` and another uses `function mapping(xi, eta)`, right? That doesn't happen right now, but I got this yesterday after experimenting with the tests.

Reply:
Yeah... Could you please open an issue about that? We should fix this stuff in the tests once and for all...

Reply:
Regarding the `refine_fn` above, in the [Julia docs](https://docs.julialang.org/en/v1/base/c/#Base.@cfunction), it says
> And that these arguments will be evaluated in global scope during compile-time (not deferred until runtime). Adding a '$' in front of the function argument changes this to instead create a runtime closure over the local variable callable (this is not supported on all architectures).

Should we still do this, even though it's "not supported on all architectures"?

Reply:
Let's leave this as is for this PR and discuss in #623.